### ISSUE TITLE
Ghost 3.0 Crashes on New Deploy w Subdirectory

### ISSUE THREAD

**Issue by bhellman1** [SNathJr/ghost-on-heroku#21]
Hello, I just discover this repo (thank you!)..

I'm attempting to launch a new/clean deploy to heroku using the "Deploy to Heroku" button seen here: https://elements.heroku.com/buttons/snathjr/ghost-on-heroku 

After deploying setting `APP_PUBLIC_URL ` to: https://mysite.io/blog

I then load:

https://mysitehere.herokuapp.com/ghost

And the app crashes (see the error log below) --- Is anything else seeing this and/or have ideas on what the issue could be? Thanks

```
Starting process with command `npm start --production
ghost-on-heroku@3.0.0 start /app
node server.js
Ghost is running in production...

/app/node_modules/ghost/core/server/services/auth/session/middleware.js:62
         req.session.user_id = req.user.id;
                             ^
TypeError: Cannot set property 'user_id' of undefined
    at /app/node_modules/ghost/core/server/services/auth/session/middleware.js:62:29
    at session (/app/node_modules/express-session/index.js:196:71)
    at getSession (/app/node_modules/ghost/core/server/services/auth/session/middleware.js:48:12)
   at Object.createSession (/app/node_modules/ghost/core/server/services/auth/session/middleware.js:52:5)
    at /app/node_modules/ghost/core/server/api/canary/session.js:35:34
    at /app/node_modules/express-brute/index.js:60:9
    at process._tickCallback (internal/process/next_tick.js:61:11)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! ghost-on-heroku@3.0.0 start: `node server.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the ghost-on-heroku@3.0.0 start script.




```

**Comment by SNathJr** [comment]
Hi @bhellman1 

This looks like it is a problem with your choice of `APP_PUBLIC_URL`. You can use your own sub domain, but you can NOT select a sub-dir as a app's URL.

Try something like https://blog.mysite.io as the Public URL and then set your CNAME accordingly.

Hope it helps!

**Comment by bhellman1** [comment]
@SNathJr I really appreciate you taking the time to reply.

I'm trying to install ghost as a subdirectory: Example: http://example.com/blog
Ghost appears to support this:
- https://ghost.org/faq/can-i-run-ghost-from-a-subdirectory/
- https://ghostboard.io/blog/how-install-ghost-in-subfolder/

Is the issue that this `ghost-on-heroku` does not support subdirectories? Or perhaps this repo has a different way of supporting a subdirectory ghost install?

Thanks

**Comment by bhellman1** [comment]
FYI, I was able to get it working by using the code from a previous PR..

Updating server.js to: 

```
var ghost = require("ghost");
var express = require("express");
var urlService = require("./node_modules/ghost/core/frontend/services/url");
var parentApp = express();

// Run a single Ghost process
ghost()
  .then(function(ghostServer) {
    ////////////////////////////////////////////////////////////////
    // this is what you need to get subdirectories working properly!
    // e.g. https://www.website.com/blog
    parentApp.use(urlService.utils.getSubdir(), ghostServer.rootApp);
    ghostServer.start(parentApp);
    ////////////////////////////////////////////////////////////////
  })
  .catch(error => {
    console.error(`Ghost server error: ${error.message} ${error.stack}`);
    process.exit(1);
  });
```

**Comment by marksbertrand** [comment]
@bhellman1 I ran into the same problem you did hosting on a subdirectory. I grabbed your server.js code but I'm getting the following error when trying to load the blog in the browser. "Cannot GET /"

Any additional steps you took besides updating the server.js file and deploying?

**Comment by desi** [comment]
@bhellman1 and @marksbertrand I too am getting the "Cannot GET /" error. Any idea what might be the issue? Any help would be greatly appreciated.

**Comment by SNathJr** [comment]
@desi @marksbertrand I will look into this issue more closely and will either create a branch or will update master with this.

Thanks!

**Comment by desi** [comment]
@SNathJr Thank you so very much! I have been banging my head trying to figure out what is happening for the last 24 hours or so. I am still trying to figure it out so if I discover anything I will also post here.

**Comment by desi** [comment]
@SNathJr I'm not sure if this is helpful but I wanted to just point out that the original thing I am trying to do is login to the ghost admin on a subdirectory deployed ghost blog, deployed to Heroku. If you don't make any changes to server.js it gives the initial error that @bhellman1 reported. If you copy the code he listed in the above comment into server.js it gives the "Cannot GET /" error. This is all against the latest ghost-on-heroku and ghost 3.2.0

**Comment by desi** [comment]
@SNathJr I am sorry to ask but do you have a time frame from when you will look at this? I am trying to figure out if I should roll back to Ghost 1.X or 2.X or whether I should wait.

**Comment by SNathJr** [comment]
@desi Sorry for the delayed reply, I was a bit busy past couple of days. It was a simple fix, however. When you have named your app url as mysite.io/blog. you should access your blog as appname.herokuapp.com/blog. That is your home URL. Also make sure to use the subdir-fix branch for deploying the app.

If this works for you as well, please put a comment so that I will be able to close this issue.

Cheers!
SNathJr

**Comment by desi** [comment]
@SNathJr Thank you so much! The advice for app url plus the branch, plus a minor change to my nginx conf seems to have done the trick! Thank you so very much! 👏

**Comment by idodekerobo** [comment]
I'm having a bit of trouble getting a subdomain to work. It seems adjacent to this issue? I'm using Cloudfare for my DNS management.


I changed the APP_PUBLIC_URL: https://blog.mydomain.me/.
It doesn't work by just changing the APP_PUBLIC_URL. So I've tried to go through the standard custom domain process for Heroku and add the CNAME record to Cloudfare for the subdomain (blog.mydomain.me/) and it still does not work.

I'm also using the root domain for another Heroku app mydomain.me/ and its working. Could this be due to that?
